File: app/src/main/java/ru/strcss/projects/moneycalc/moneycalcandroid/api/MoneyCalcClient.java
package ru.strcss.projects.moneycalc.moneycalcandroid.api;

import java.util.List;

import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import ru.strcss.projects.moneycalc.moneycalcdto.dto.Credentials;
import ru.strcss.projects.moneycalc.moneycalcdto.dto.MoneyCalcRs;
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.settings.SpendingSectionUpdateContainer;
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.statistics.FinanceSummaryFilterLegacy;
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.transactions.TransactionUpdateContainerLegacy;
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.transactions.TransactionsSearchFilterLegacy;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.Access;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.FinanceSummaryBySection;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.SettingsLegacy;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.SpendingSection;
import ru.strcss.projects.moneycalc.moneycalcdto.entities.TransactionLegacy;
import rx.Observable;

public interface MoneyCalcClient {
    @POST("/api/registration/register")
    Observable<MoneyCalcRs<Void>> registerPerson(@Body Credentials credentials);

    @POST("/login")
    Observable<Response<Void>> login(@Body Access access);

    @GET("/api/settings")
    Observable<MoneyCalcRs<SettingsLegacy>> getSettings(@Header("Authorization") String token);

    @PUT("/api/settings")
    Observable<MoneyCalcRs<SettingsLegacy>> updateSettings(@Header("Authorization") String token,
                                                           @Body SettingsLegacy settings);

    /**
     * Spending Sections
     */

    @GET("/api/spendingSections")
    Observable<MoneyCalcRs<List<SpendingSection>>> getSpendingSections(@Header("Authorization") String token);

    @POST("/api/spendingSections")
    Observable<MoneyCalcRs<List<SpendingSection>>> addSpendingSection(@Header("Authorization") String token,
                                                                      @Body SpendingSection spendingSection);

    @PUT("/api/spendingSections")
    Observable<MoneyCalcRs<List<SpendingSection>>> updateSpendingSection(@Header("Authorization") String token,
                                                                         @Body SpendingSectionUpdateContainer updateContainer);

    @DELETE("/api/spendingSections/{sectionId}")
    Observable<MoneyCalcRs<List<SpendingSection>>> deleteSpendingSection(@Header("Authorization") String token,
                                                                         Integer sectionId);

    /**
     * Statistics
     */
    @GET("/api/stats/summaryBySection")
    Observable<MoneyCalcRs<List<FinanceSummaryBySection>>> getFinanceSummaryBySection(@Header("Authorization") String token);

    @POST("/api/stats/summaryBySection/getFiltered")
    Observable<MoneyCalcRs<List<FinanceSummaryBySection>>> getFinanceSummaryBySection(@Header("Authorization") String token,
                                                                                      @Body FinanceSummaryFilterLegacy getContainer);

    /**
     * Transactions
     */
    @POST("/api/transactions")
    Observable<MoneyCalcRs<TransactionLegacy>> addTransaction(@Header("Authorization") String token,
                                                              @Body TransactionLegacy transaction);

    @GET("/api/transactions")
    Observable<MoneyCalcRs<List<TransactionLegacy>>> getTransactions(@Header("Authorization") String token);

    @POST("/api/transactions/getFiltered")
    Observable<MoneyCalcRs<List<TransactionLegacy>>> getTransactions(@Header("Authorization") String token,
                                                                     @Body TransactionsSearchFilterLegacy container);

    @PUT("/api/transactions")
    Observable<MoneyCalcRs<TransactionLegacy>> updateTransaction(@Header("Authorization") String token,
                                                                 @Body TransactionUpdateContainerLegacy transactionUpdateContainer);

    @DELETE("/api/transactions/{transactionId}")
    Observable<MoneyCalcRs<Void>> deleteTransaction(@Header("Authorization") String token,
                                                    Integer transactionId);
}

File: app/src/main/java/ru/strcss/projects/moneycalc/moneycalcandroid/api/MoneyCalcClient.kt
package ru.strcss.projects.moneycalc.moneycalcandroid.api

import retrofit2.Response
import retrofit2.http.*
import ru.strcss.projects.moneycalc.moneycalcdto.dto.Credentials
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.ItemsContainer
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.settings.SpendingSectionUpdateContainer
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.spendingsections.SpendingSectionsSearchRs
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.statistics.StatisticsFilter
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.transactions.TransactionUpdateContainerLegacy
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.transactions.TransactionsSearchFilterLegacy
import ru.strcss.projects.moneycalc.moneycalcdto.dto.crudcontainers.transactions.TransactionsSearchLegacyRs
import ru.strcss.projects.moneycalc.moneycalcdto.entities.*
import ru.strcss.projects.moneycalc.moneycalcdto.entities.statistics.SumByDate
import ru.strcss.projects.moneycalc.moneycalcdto.entities.statistics.SumByDateSection
import ru.strcss.projects.moneycalc.moneycalcdto.entities.statistics.SumBySection
import ru.strcss.projects.moneycalc.moneycalcdto.entities.statistics.SummaryBySection
import rx.Observable

interface MoneyCalcClient {
    @POST("/api/registration/register")
    fun registerPerson(@Body credentials: Credentials): Observable<Person>

    @POST("/login")
    fun login(@Body access: Access): Observable<Response<Void>>

    @GET("/api/settings")
    fun getSettings(@Header("Authorization") token: String?): Observable<SettingsLegacy>

    @PUT("/api/settings")
    fun updateSettings(@Header("Authorization") token: String?,
                       @Body settings: SettingsLegacy): Observable<SettingsLegacy>

    /**
     * Spending Sections
     */

    @GET("/api/spendingSections")
    fun getSpendingSections(@Header("Authorization") token: String?): Observable<SpendingSectionsSearchRs>

    @POST("/api/spendingSections")
    fun addSpendingSection(@Header("Authorization") token: String?,
                           @Body spendingSection: SpendingSection): Observable<SpendingSectionsSearchRs>

    @PUT("/api/spendingSections")
    fun updateSpendingSection(@Header("Authorization") token: String?,
                              @Body updateContainer: SpendingSectionUpdateContainer): Observable<SpendingSectionsSearchRs>

    @DELETE("/api/spendingSections/{sectionId}")
    fun deleteSpendingSection(@Header("Authorization") token: String?,
                              @Path("sectionId") sectionId: Int?): Observable<SpendingSectionsSearchRs>

    /**
     * Statistics By Section
     */
    @GET("/api/stats/bySection/summary")
    fun getStatsBySectionSummary(@Header("Authorization") token: String?): Observable<ItemsContainer<SummaryBySection>>

    @POST("/api/stats/bySection/sum")
    fun getStatsBySectionSum(@Header("Authorization") token: String?,
                             @Body getContainer: StatisticsFilter): Observable<ItemsContainer<SumBySection>>

    /**
     * Statistics By Date
     */
    @POST("/api/stats/byDate/sum")
    fun getStatsByDateSum(@Header("Authorization") token: String?,
                          @Body getContainer: StatisticsFilter): Observable<ItemsContainer<SumByDate>>

    @POST("/api/stats/byDate/sumBySection")
    fun getStatsByDateSumBySection(@Header("Authorization") token: String?,
                                   @Body getContainer: StatisticsFilter): Observable<ItemsContainer<SumByDateSection>>

    /**
     * Transactions
     */
    @GET("/api/transactions")
    fun getTransactions(@Header("Authorization") token: String?): Observable<TransactionsSearchLegacyRs>

    @POST("/api/transactions")
    fun addTransaction(@Header("Authorization") token: String?,
                       @Body transaction: TransactionLegacy): Observable<TransactionLegacy>

    @POST("/api/transactions/getFiltered")
    fun getTransactions(@Header("Authorization") token: String?,
                        @Body container: TransactionsSearchFilterLegacy): Observable<TransactionsSearchLegacyRs>

    @PUT("/api/transactions")
    fun updateTransaction(@Header("Authorization") token: String?,
                          @Body transactionUpdateContainer: TransactionUpdateContainerLegacy): Observable<TransactionLegacy>

    @DELETE("/api/transactions/{transactionId}")
    fun deleteTransaction(@Header("Authorization") token: String?,
                          @Path("transactionId") transactionId: Int?): Observable<Void>
}


